MNT: `io.ascii` deprecate and rename / remove arguments

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to formally deprecate a number of keyword arguments that have non-PEP8-compliant names and/or functionality that has been superceded by a new preferred API. In summary:

``read()``:
- ``Reader`` will be removed. Instead supply the equivalent ``format`` argument.
- ``Inputter`` has been renamed to ``inputter_cls``.
- ``Outputter`` has been renamed to ``outputter_cls``.

``get_reader()``:
- ``Reader`` has been renamed to ``reader_cls``.
- ``Inputter`` has been renamed to ``inputter_cls``.
- ``Outputter`` has been renamed to ``outputter_cls``.

``write()``:
- ``Writer`` will be removed. Instead supply the equivalent ``format`` argument.

``get_writer()``:
- ``Writer`` has been renamed to ``writer_cls``.


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #8567

pllim commented:
Do you think this will break a lot of downstream code? I see a lot of API changes but I cannot tell what is internal use only and what isn't. We might want to consider:

* Sending out a memo to astropy-dev about this PR.
* Repeating that memo in Astropy Slack `#general` and mirror to relevant sub-channels.
* If no objection, merge this soon so people have a chance to change this downstream early in the dev cycle.

taldcroft commented:
It doesn't hurt to mention in this in a few places and I'll do that. However, I am not very worried about a lot of code breakage. The `Reader=<reader_class>` and `Writer=<writer_class>` interfaces are not in any actual code examples and it has been this way for many years. Likewise using the `get_reader()` and `get_writer()` functions should be rare for most users.

I think that part of why this looks so impacting is that a lot of the tests were written prior to updating the interface (to use `format` instead of `Reader`/`Writer`). And of course a lot of the internal implementation code uses these functions resulting in a lot of diffs.

pllim commented:
`specutils`, for one, registers their own reader/writer at https://github.com/astropy/specutils/blob/main/specutils/io/default_loaders/ascii.py though at a glance look like they will be okay?

taldcroft commented:
> specutils, for one, registers their own reader/writer at https://github.com/astropy/specutils/blob/main/specutils/io/default_loaders/ascii.py though at a glance look like they will be okay?

Yup, it reads a table using `Table.read(.., format=...)`, which is the desired API and is unaffected.

hamogu commented:
@taldcroft In my view, our "simpler landing page" is https://www.astropy.org/ 
From there you get to the docs. Sure, https://www.astropy.org/ might also profit from a re-design, but as it is I'm not sure that it would be beneficial to have a landing page there, which then gets you to a simple landing page for the docs, which then gets you to a section heading. Seems like too many levels of nesting. Not sure what do do about it, just commenting with my opinion.

hamogu commented:
Upp,s pasted comment on the wrong issue.... Sorry.

taldcroft commented:
@hamogu - sorry for the late update, but I just noticed this PR hanging out. I rebased, fixed some issues and addressed your review comment for additional tests. It looks like tests are passing though a few combinations are still running. I think we're ready to finally deprecate those ancient `asciitable` names!

hamogu reviewed (approved):
Long live `asciitable`!

pllim commented:
What is still deprecated and need to be removed in 7.0 or later?

taldcroft commented:
@pllim - If I understand the question, in the description there are 8 function kwargs that are now formally deprecated and they should all be removed in 7.0.

pllim commented:
@taldcroft , can you please open a follow up issue and list the 8 things slated for removal? Hard to tell from all the diff here. Thanks!

taldcroft commented:
@pllim will #15536 do the trick so we remember? Or maybe I need a calendar reminder for a year from now?

pllim commented:
Haha, calendar reminder is not necessary. Thanks for opening the issue! 😸
